lxd/storage/drivers/volume.go

```diff
@@ -314,7 +314,7 @@ func (v Volume) NewVMBlockFilesystemVolume() Volume {
 	}
 
 	// VM Block filesystems are a fixed size.
-	newConf["size"] = vmBlockFilesystemSize
+	newConf["size"] = DefaultVMBlockFilesystemSize
```

Comment:
@stgraber I think this is going to cause problems, for example when copying the instance, I think its likely the vm state volume will get re-created at the default size. Perhaps `NewVMBlockFilesystemVolume` needs to be updated to take a size argument, and if empty then falls back to the default (which would then avoid the duplicate logic in  `SetInstanceQuota`

Reply:
Fixed everything except this one. I'm not sure how changing `NewVMBlockFilesystemVolume` helps us with the migration case. Don't we just need to add a new migration field to carry the size of the fs volume?

Reply:
@stgraber so the `size` property of a `Volume` is used by the storage driver to set its initial size when creating a volume (rather than updating it). E.g. https://github.com/lxc/lxd/blob/master/lxd/storage/drivers/driver_zfs_volumes.go#L139

This can be, for example, when creating a volume from an instance with profile config, or doing migrations.

So we need to ensure that the instance config value `size.state` gets propagated to the `size` property for the `Volume` that is returned from `NewVMBlockFilesystemVolume`, otherwise it will be created using the default size.

Given that anywhere we are operating on a VM filesystem volume, we expect to have to derive that from the associated VM block volume (which is the volume type passed into the storage driver), I think doing this will work:

1. Modify `instanceRootVolumeConfig` function (which already "munges" the DB storage volume config with the instance disk device's config into one set of configs for the storage layer) to copy the instance's `size.state` into the returned config as `size.state`.
2. That in turn will be stored in the VM block `Volume` config passed into the driver layer, so that when we call `vol.NewVMBlockFilesystemVolume()` we can update `NewVMBlockFilesystemVolume` to detect the presence of `size.state` in the VM block volume's config and copy that into the returned FS volume's `size` property. Or fall back to the default as we do now.

That way any operations that we do on the subsequently created filesystem volume returned from `NewVMBlockFilesystemVolume` will have the correct `size` property set in its config.

Reply:
Ok, I made a small change to the storage package now which sorts out the creation of new instances from a profile and also handles the local pool migration properly as far as I can tell.

Reply:
@tomponline this should be good enough for now. When we add live migration we'll want to extend the migration protobuf protocol to include the size so we can handle those remaining cases.

Reply:
@stgraber looks good. I'll wait for tests to pass before merging though.

As for migration, as we send the instance config via migration already, there's a good chance we won't need to handle that separately as it'll be handled now by `instanceRootVolumeConfig`.

Reply:
@tomponline all green now